client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -309,12 +311,17 @@ private StreamCut computeEndStreamCut(Stream stream, Map<Segment, Long> endSegme
         }
         return unread.thenApply(unreadVal -> {
             long totalLength = 0;
+            DelegationTokenProvider tokenProvider = null;
             for (Segment s : unreadVal.getSegments()) {
                 if (endPositions.containsKey(s)) {
                     totalLength += endPositions.get(s);
                 } else {
+                    if (tokenProvider == null) {
+                        tokenProvider = DelegationTokenProviderFactory.create(
```

Comment:
I've added a `populateToken()` method to the `DelegationTokenProvider` interface that can be used to populate tokens retrieved later through piggybacking into the provider. However, that doesn't seem useful here as there is no instance variable for holding the provider. I didn't want to modify the logic of this class in a big way, so I didn't try introducing a new `DelegationTokenProvider` instance variable in this class. Could we live with this?

Reply:
At the top of the loop instead of `DelegationTokenProvider tokenProvider = null;`
You can do `DelegationTokenProvider tokenProvider = DelegationTokenProviderFactory.create(...)`;
Then here you don't need this `if`. You can just call `populateToken(unreadVal.getDelegationToken())`

Reply:
We also need both `scope name` and `stream name` to be able to initialize the `DelegationTokenProvider` here. The scope name is present as an instance variable, but not the stream name. So, I haven't been able to make this change, and I am using a `Segment` object for the purpose, which contains both scope name and stream name.  Is there a way I can obtain the stream name without going through any hoops?

Reply:
Yes: `fromStreamCut.asImpl().getStream()`